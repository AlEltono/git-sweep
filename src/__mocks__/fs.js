const path = require('path');

const fs = jest.genMockFromModule('fs');

// This is a custom function that our tests can use during setup to specify
// what the files on the "mock" filesystem should look like when any of the
// `fs` APIs are used.
let mockFiles = Object.create(null);
function __setMockFiles(newMockFiles) {
  mockFiles = Object.create(newMockFiles);
}

function readFile(path, encoding, cb) {
  if (mockFiles[path]) {
    cb(null, mockFiles[path]);
  } else {
    cb(new Error());
  }
}

fs.__setMockFiles = __setMockFiles;
fs.readFile = readFile;

module.exports = fs;
