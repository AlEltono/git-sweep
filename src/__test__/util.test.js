import path from 'path';
import moment from 'moment';

jest.mock('fs');

describe('getConfiguredIgnoresIfExist', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('rejects when `repoPath` is missing', async () => {
    const util = require('../util');

    try {
      await util.getConfiguredIgnoresIfExist()
    } catch (e) {
      expect(e.message).toEqual('`repoPath` is required');
    }
  });

  it('attempts to read ".gitsweepignore" file', () => {
    const fs = require('fs');
    const util = require('../util');
    fs.readFile = jest.fn();

    util.getConfiguredIgnoresIfExist('.');

    expect(fs.readFile.mock.calls[0][0]).toBe(path.join('.', '.gitsweepignore'));
  });

  it('resolves with empty array if file not found', async () => {
    const util = require('../util');
    const result = await util.getConfiguredIgnoresIfExist('.');

    expect(result).toEqual([]);
  });

  it('resolves with a list of ignored branches', async () => {
    require('fs').__setMockFiles({
      [path.join('.', '.gitsweepignore')]: `
        origin/master
        origin/dev
      `
    });
    const util = require('../util');

    const result = await util.getConfiguredIgnoresIfExist('.');

    expect(result).toEqual(['origin/master', 'origin/dev']);
  });
});

describe('getCutoffMoment', () => {
  let currentMoment;

  beforeEach(() => {
    currentMoment = moment('2017-01-01T00:00:00+0000');
  });

  it('recognizes `y` token', () => {
    const util = require('../util');

    const cutoffMoment = util.getCutoffMoment(currentMoment, '1y');

    expect(cutoffMoment).toMatchSnapshot();
  });

  it('recognizes `m` token', () => {
    const util = require('../util');

    const cutoffMoment = util.getCutoffMoment(currentMoment, '1m');

    expect(cutoffMoment).toMatchSnapshot();
  });

  it('recognizes `d` token', () => {
    const util = require('../util');

    const cutoffMoment = util.getCutoffMoment(currentMoment, '1d');

    expect(cutoffMoment).toMatchSnapshot();
  });

  it('gives correct cutoff moment for various combinations', () => {
    const util = require('../util');

    expect(util.getCutoffMoment(currentMoment, '1y2m3d')).toMatchSnapshot();
    expect(util.getCutoffMoment(currentMoment, '1y3d')).toMatchSnapshot();
    expect(util.getCutoffMoment(currentMoment, '2m3d')).toMatchSnapshot();
    expect(util.getCutoffMoment(currentMoment, '1y2m')).toMatchSnapshot();
  });
});

describe('isIgnored', () => {
  it('works when ignoreList is empty', () => {
    const util = require('../util');

    expect(util.isIgnored('refs/remotes/origin/dev', [])).toBe(false);
  });

  it('checks whether a ref is in ignoredList', () => {
    const util = require('../util');

    const list = ['origin/dev'];

    expect(util.isIgnored('refs/remotes/origin/dev', list)).toBe(true);
    expect(util.isIgnored('refs/remotes/origin/feature', list)).toBe(false);
  });
});
