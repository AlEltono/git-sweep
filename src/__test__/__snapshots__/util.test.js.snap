// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCutoffMoment gives correct cutoff moment for various combinations 1`] = `"2015-10-29T00:00:00.000Z"`;

exports[`getCutoffMoment gives correct cutoff moment for various combinations 2`] = `"2015-12-29T00:00:00.000Z"`;

exports[`getCutoffMoment gives correct cutoff moment for various combinations 3`] = `"2016-10-29T00:00:00.000Z"`;

exports[`getCutoffMoment gives correct cutoff moment for various combinations 4`] = `"2015-11-01T00:00:00.000Z"`;

exports[`getCutoffMoment recognizes \`d\` token 1`] = `"2016-12-31T00:00:00.000Z"`;

exports[`getCutoffMoment recognizes \`m\` token 1`] = `"2016-12-01T00:00:00.000Z"`;

exports[`getCutoffMoment recognizes \`y\` token 1`] = `"2016-01-01T00:00:00.000Z"`;
